go/kbfs/simplefs/archive.go

```diff
@@ -0,0 +1,876 @@
+	}
+}
+
+type sha256TeeReader struct {
+	inner          io.Reader
+	innerTeeReader io.Reader
+	h              hash.Hash
+}
+
+var _ io.Reader = (*sha256TeeReader)(nil)
+
+// Read implements the io.Reader interface.
+func (r *sha256TeeReader) Read(p []byte) (n int, err error) {
+	return r.innerTeeReader.Read(p)
+}
+
+func (r *sha256TeeReader) getSum() []byte {
+	return r.h.Sum(nil)
+}
+
+func newSHA256TeeReader(inner io.Reader) (r *sha256TeeReader) {
+	r = &sha256TeeReader{
+		inner: inner,
+		h:     sha256.New(),
+	}
+	r.innerTeeReader = io.TeeReader(r.inner, r.h)
+	return r
+}
+
+func (m *archiveManager) copyFileFromBeginning(ctx context.Context,
+	srcDirFS billy.Filesystem, entryPathWithinJob string,
+	localPath string, mode os.FileMode) (sha256Sum []byte, err error) {
+	m.simpleFS.log.CDebugf(ctx, "+ copyFileFromBeginning %s", entryPathWithinJob)
+	defer func() { m.simpleFS.log.CDebugf(ctx, "- copyFileFromBeginning %s err: %v", entryPathWithinJob, err) }()
+
+	src, err := srcDirFS.Open(entryPathWithinJob)
+	if err != nil {
+		return nil, fmt.Errorf("srcDirFS.Open(%s) error: %v", entryPathWithinJob, err)
+	}
+	defer src.Close()
+
+	dst, err := os.OpenFile(localPath, os.O_APPEND|os.O_WRONLY|os.O_CREATE, mode)
+	if err != nil {
+		return nil, fmt.Errorf("os.OpenFile(%s) error: %v", localPath, err)
+	}
+	defer dst.Close()
+
+	teeReader := newSHA256TeeReader(src)
+
+loopCopy:
+	for {
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		default:
+		}
+		_, err := io.CopyN(dst, teeReader, 64*1024)
+		switch err {
+		case nil:
+		case io.EOF:
+			break loopCopy
+		default:
+			return nil, fmt.Errorf("[%s] io.CopyN error: %v", entryPathWithinJob, err)
+		}
+	}
+
+	// We didn't continue from a previously interrupted copy, so don't
+	// bother verifying the sha256sum and just return it.
+	return teeReader.getSum(), nil
+}
+
+func (m *archiveManager) copyFilePickupPrevious(ctx context.Context,
+	srcDirFS billy.Filesystem, entryPathWithinJob string,
+	localPath string, srcSeekOffset int64, mode os.FileMode) (sha256Sum []byte, err error) {
+	m.simpleFS.log.CDebugf(ctx, "+ copyFilePickupPrevious %s", entryPathWithinJob)
+	defer func() { m.simpleFS.log.CDebugf(ctx, "- copyFilePickupPrevious %s err: %v", entryPathWithinJob, err) }()
+
+	src, err := srcDirFS.Open(entryPathWithinJob)
+	if err != nil {
+		return nil, fmt.Errorf("srcDirFS.Open(%s) error: %v", entryPathWithinJob, err)
+	}
+	defer src.Close()
+
+	_, err = src.Seek(srcSeekOffset, io.SeekStart)
+	if err != nil {
+		return nil, fmt.Errorf("[%s] src.Seek error: %v", entryPathWithinJob, err)
+	}
+
+	// Copy the file.
+	if err = func() error {
+		dst, err := os.OpenFile(localPath, os.O_APPEND|os.O_WRONLY|os.O_CREATE, mode)
+		if err != nil {
+			return fmt.Errorf("os.OpenFile(%s) error: %v", localPath, err)
+		}
+		defer dst.Close()
+
+	loopCopy:
+		for {
+			select {
+			case <-ctx.Done():
+				return ctx.Err()
+			default:
+			}
+			_, err := io.CopyN(dst, src, 64*1024)
```

Comment:
I assume you're not using `io.Copy()` here because you want it to be responsive to cancels?  Might be worth a comment.

Reply:
I actually just copied the CopyN over from the old copy code but I guess that was the purpose there as well! I'll comment

Reply:
(I ended up just naming the new function `ctxAwareCopy` to make it clear, insterad of a comment)